### PR TITLE
add error rate metrics for kafka

### PR DESCRIPTION
**Pull request by akaspin**
[trello](https://trello.com/c/LbNrPTxT)

### CODE REVIEW

File: kafka_tracker.go [hunk 1]
```diff
@@ -4,12 +4,18 @@ import (
...
 	"github.com/Shopify/sarama"
+	"github.com/rcrowley/go-metrics"
 	"github.com/remerge/cue"
 )
 
 // KafkaTracker is a tracker that sends messages to Apache Kafka.
 type KafkaTracker struct {
 	BaseTracker
+	metrics struct {
+		// TODO (Aleksandr Dorofeev): Provide registry to Sarama
```
Comment (itszootime) [inline kafka_tracker.go]: Let's try not to leave a `TODO` in the code 🙂

File: kafka_tracker.go [hunk 2]
```diff
@@ -39,15 +45,27 @@ func NewKafkaTracker(
...
 
 	config.Producer.Return.Successes = false
-	config.Producer.Return.Errors = false
+	config.Producer.Return.Errors = true
 	config.Producer.RequiredAcks = sarama.NoResponse
 	config.Producer.Compression = sarama.CompressionSnappy
 	config.ChannelBufferSize = 131072
 
+	t.metrics.fastErrorRate = metrics.GetOrRegisterMeter(
+		"tracker,type=fast kafka_send_error_rate",
```
Comment (itszootime) [inline kafka_tracker.go]: `kafka_produce_error_rate` would be a bit more precise

File: kafka_tracker.go [hunk 3]
```diff
@@ -39,15 +45,27 @@ func NewKafkaTracker(
 	)
 
 	config.Producer.Return.Successes = false
-	config.Producer.Return.Errors = false
+	config.Producer.Return.Errors = true
 	config.Producer.RequiredAcks = sarama.NoResponse
 	config.Producer.Compression = sarama.CompressionSnappy
 	config.ChannelBufferSize = 131072
 
+	t.metrics.fastErrorRate = metrics.GetOrRegisterMeter(
```
Comment (itszootime) [inline kafka_tracker.go]: I don't think we need to break this line, it's not that long

File: kafka_tracker.go [hunk 4]
```diff
@@ -69,6 +87,9 @@ func NewKafkaTracker(
 		_ = log.Error(t.kafka.fast.Close(), "failed to close fast producer")
 		return nil, err
 	}
+	t.metrics.safeErrorRate = metrics.GetOrRegisterMeter(
```
Comment (itszootime) [inline kafka_tracker.go]: Same comments here as for the `fastErrorRate`

File: kafka_tracker.go [hunk 5]
```diff
@@ -39,15 +45,27 @@ func NewKafkaTracker(
...
+		nil)
+
 	t.kafka.fast, err = sarama.NewAsyncProducer(brokers, config)
 	if err != nil {
 		return nil, err
 	}
+	go func() {
+		for fastErr := range t.kafka.fast.Errors() {
+			t.metrics.fastErrorRate.Mark(1)
+			_ = log.WithFields(cue.Fields{
```
Comment (itszootime) [inline kafka_tracker.go]: This will send a Rollbar alert for every produce error. If we want to keep the logging, we'll have to use `INFO` level

**Comment by itszootime** [comment]
Won't we also need a public way to access the error rate for the bidder health checks?

File: kafka_tracker.go (hunk 1 above)
Comment (akaspin) [reply kafka_tracker.go]: OK. I'll convert this into Trello card.

File: kafka_tracker.go (hunk 2 above)
Comment (akaspin) [reply kafka_tracker.go]: OK.

File: kafka_tracker.go (hunk 3 above)
Comment (akaspin) [reply kafka_tracker.go]: Linter swears :-)

**Comment by akaspin** [comment]
> Won't we also need a public way to access the error rate for the bidder health checks?

@itszootime In next PR. The good way is not to expose metrics but provide separate API method.

File: kafka_tracker.go (hunk 5 above)
Comment (itszootime) [reply kafka_tracker.go]: [`WARN` would also be fine](https://github.com/remerge/go-service/blob/master/service.go#L206)

Comment (akaspin) [reply kafka_tracker.go]: Nice.

**Pull request merged**